doc/src/explanation/best-practices.md

```diff
@@ -0,0 +1,1329 @@
+to just start with rational numbers in the first place, if you can):
+
+```py
+>>> from sympy import nsimplify
+>>> Rational(0.7)
+3152519739159347/4503599627370496
+>>> nsimplify(0.7)
+7/10
+```
+
+### Avoid `simplify()`
+
+{func}`~.simplify` (not to be confused with {func}`~.sympify`) is designed as
+a general purpose heuristic. It tries various simplification algorithms on the
+input expression and returns the result that seems the "simplest" based on
+some metric.
+
+`simplify()` is perfectly fine for interactive use, where you just want SymPy
+to do whatever it can to an expression. However, in programmatic usage, it's
+better to avoid `simplify()` and use more [targeted simplification
+functions](simplify-docs) instead (e.g., {func}`~.cancel`, {func}`~.expand`,
+or {func}`~.collect`).
+
+There are a few reasons why this is generally preferred:
+
+- Due to its heuristical nature, `simplify()` can potentially be slow, since
+  it tries a lot of different approaches to try to find the best
+  simplification.
+
+- There are no guarantees about what form an expression will have after being
+  passed through `simplify()`. It may actually end up "less simple" by
+  whatever metric you were hoping for. To contrast, targeted simplification
+  functions are very specific about what behaviors they have and what they
+  guarantee about the output. For example,
+
+  - {func}`~.factor` will always factor a polynomial into irreducible factors.
+
+  - {func}`~.cancel` will always convert a rational function into the form
+    $p/q$ where $p$ and $q$ are expanded polynomials with no common factors.
+
+  The documentation for each function describes exactly what behavior it will
+  have on the input expression.
+
+- A targeted simplification will not do something unexpected if the expression
+  contains an unexpected form, or an unexpected subexpression. This is
+  especially the case if simplification functions are applied with `deep=False`
+  to only apply the simplification to the top-level expression.
+
+Some other simplification functions are heuristical in nature, and care should
+be taken with them as well. For example, the {func}`~.trigsimp` function is a
+heuristic targeted to trigonometric functions, but the routines in the
+{mod}`sympy.simplify.fu` submodule allow applying specific trigonometric
+identities.
+
+The [simplify section of the tutorial](tutorial-simplify) and the
+[simplify module reference](../modules/simplify/simplify.rst) list the various
+targeted simplification functions.
+
+In some cases, you may know exactly what simplification operations you wish to
+apply to an expression, but there may not be an exact set of simplification
+functions that do them. When this happens, you can create your own targeted
+simplification using {meth}`~sympy.core.basic.Basic.replace`, or in general, manually using
+[advanced expression manipulation](tutorial-manipulation).
+
+<!-- TODO: Add an example -->
```

Comment:
Can anyone suggest a good example to demonstrate why using `simplify()` should be avoided (especially in programmatic contexts) over more targeted simplification functions like `cancel()`?

Reply:
I think @oscarbenjamin mentioned one recently that had to do with hyperbolics appearing that made some roots more complicated. Trig functions without targeted (Fu-like) transpormation are notorious for getting unwieldy. Also, even something simple like 
```python
>>> exp(x)/2 + exp(-x)/2
exp(x)/2 + exp(-x)/2
>>> simplify(_)
cosh(x)
```
is simpler but one might not have wanted to convert to hyperbolics.

Reply:
The example from https://github.com/sympy/sympy/pull/25458#issuecomment-1662397968 is a good one:
```python
In [10]: M = Matrix([[0, -t*cos(t) + sin(t)], [-t*cos(t) + sin(t), 0]])

In [11]: M
Out[11]: 
⎡        0           -t⋅cos(t) + sin(t)⎤
⎢                                      ⎥
⎣-t⋅cos(t) + sin(t)          0         ⎦

In [12]: p = M.charpoly(z).as_expr()

In [13]: p
Out[13]: 
 2                      2
z  - (t⋅cos(t) - sin(t)) 

In [14]: roots(p, z)
Out[14]: 
⎧    ___________________________________        ___________________________________   ⎫
⎨   ╱  2    2                      2           ╱  2    2                      2       ⎬
⎩-╲╱  t ⋅cos (t) - t⋅sin(2⋅t) + sin (t) : 1, ╲╱  t ⋅cos (t) - t⋅sin(2⋅t) + sin (t) : 1⎭
```
Why didn't roots spot that there is a simpler expression for the roots here? Where did `sin(2*t)` come from?

The answer is:
```python
In [15]: p.expand()
Out[15]: 
   2    2                           2      2   
- t ⋅cos (t) + 2⋅t⋅sin(t)⋅cos(t) + z  - sin (t)

In [16]: p.expand().simplify()
Out[16]: 
   2    2                    2      2   
- t ⋅cos (t) + t⋅sin(2⋅t) + z  - sin (t)

In [17]: p.expand().collect(z, factor)
Out[17]: 
 2                      2
z  - (t⋅cos(t) - sin(t)) 

In [18]: p.expand().simplify().collect(z, factor)
Out[18]: 
   2    2                    2      2   
- t ⋅cos (t) + t⋅sin(2⋅t) + z  - sin (t)
```
So somewhere `simplify` (or `trigsimp`) is used and it decides to replace `2*sin(t)*cos(t)` with `sin(2*t)`. Whether or not this is a good idea is very context dependent but if you intend to do algebra like `factor`, `cancel` etc then it is definitely not better to introduce new "generators" (although eliminating some generators in favour of others that are *already* present is usually good). Here the use of `simplify` turns an expression that could have been factorised into one that cannot be factorised. In turn this takes us from nice rational expressions for the roots that are free of branch cuts to horrible trig-radical expressions.

I'm not sure why `eigenvals` itself is not affected but all of these functions misuse or overuse `simplify`:
```python
In [20]: M.eigenvals()
Out[20]: {-t⋅cos(t) + sin(t): 1, t⋅cos(t) - sin(t): 1}
```
The fact that `simplify` is actually being used *inside* functions like `roots`, `eigenvals` etc means that it is not even possible to carefully manipulate your inputs into the form you want before calling those functions.

Reply:
I'll have to think about this. A potential problem with an example like this is that it's hard to not just say "well, shouldn't `simplify()` just be smarter?" (e.g., in this case, try to rewrite the expression in such a way that makes it more factorable before calling `factor()`). And indeed, it's quite possible that at some point in the future it *will* become smarter, which would completely break the example.

It would be more robust to have an example where simplify() pretty unambiguously returns a "simple" form of an expression by some measure, but you can also say by some other measure that it isn't simple.

Maybe I can manipulate this example into something like that, but it's hard because you're really just saying "`simplify()` returned an expression that wasn't the correct 'simplified' for `simplify()`". But it would be I think more instructive to show an example like "`simplify()` returned something that wasn't the correct 'simplified' for `<some other function>`".

Another idea: even just for a basic rational function, there's at least 3 "simple" ways it can be represented: as p/q with p and q expanded and no factors in common (i.e., `cancel()`), with p/q with p and q factored (`factor()`), or as a partial fraction decomposition (`apart()`). `simplify()` seems to prefer the first of these, but there's no reason in particular for that, and it might some day switch to one of the others. But the point is that there are situations where all three forms can be useful.

Reply:
I think that anything that you might write here would be potentially subject to change in future because simplify is heuristic. The point of my example is that it is a real problem that is actually caused by the use of simplify right now in the SymPy codebase. Of course we *should* try to fix that in future but it demonstrates why naive use of simplify is not always good (besides speed concerns) because it can lead to unpredictable results that ruin a larger calculation.

Reply:
I think that the problem of simplify

- Ambiguity problem:
  In the PR https://github.com/sympy/sympy/pull/25456, it is discovered that simplify is very hard to trigger and is very difficult to quantify when does it trigger.

- Abstraction problem:
  In the dependency inversion principle, high level features shouldn't depend on the low level implementation, but abstractions of it.
 `simplify` is very specific to the implementation details of computer algebra itself, than the abstract algebra. For example, I'm very skeptical about if transfer function should compute feedback result, or quaternions should compute norms with `simplify` than do bare algebra.
  This can later cause significant restriction if you want to let users of use implementations of symengine, or do other customizations.

Reply:
I think that the only production and practical usage of `simplify` should be for compilation workflow.
And similar as how compiler works, the compiled code can never be reused by humans, and can only interpreted by compilers or machines.
Similar as that, the result of simplify is never interpretable than other `simplify` or the printers.

Reply:
I don't think that `simplify` should even be used for a compilation workflow: if you want to optimise an expression for turning into compiled code then you want a specialised simplify function for that.

Basically I don't think there is any situation within SymPy itself where `simplify` should be used but it is a nice thing for users to use interactively that will likely do what they want in simple cases.

Reply:
That's actually a good point. I think in practice, more people have the issue that simplify does too little, that it doesn't happen to actually call the simplification function they are looking for.